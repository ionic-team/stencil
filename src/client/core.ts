import { AppGlobal } from '../util/interfaces';
import { Build } from '../util/build-conditionals';
import { createPlatformClient } from './platform-client';


const App: AppGlobal = (<any>window)[appNamespace] = (<any>window)[appNamespace] || {};

const plt = createPlatformClient(Context, App, window, document, publicPath, hydratedCssClass);


if (Build.es5) {
  plt.registerComponents(App.components).forEach(cmpMeta => {
    // es5 way of extending HTMLElement
    function HostElement(self: any) {
      return HTMLElement.call(this, self);
    }

    HostElement.prototype = Object.create(
      HTMLElement.prototype,
      { constructor: { value: HostElement, configurable: true } }
    );

    plt.defineComponent(cmpMeta, HostElement);
  });

} else {
  // es6 class extends HTMLElement
  plt.registerComponents(App.components).forEach(cmpMeta =>
    plt.defineComponent(cmpMeta, class extends HTMLElement {}));
}
