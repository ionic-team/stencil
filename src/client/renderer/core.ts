/**
 * Virtual DOM patching algorithm based on Snabbdom by
 * Simon Friis Vindum (@paldepind)
 * Licensed under the MIT License
 * https://github.com/snabbdom/snabbdom/blob/master/LICENSE
 *
 * Modified for Ionic's Web Component Renderer
 * /


/* global module, document, Node */
import { isArray, isDef, isUndef, isString, isStringOrNumber } from '../../util/helpers';
import { HostContentNodes, PlatformApi, ProxyElement, RendererApi, VNode, VNodeData, Key } from '../../util/interfaces';
import { vnode } from './vnode';

import { updateAttrs } from './modules/attributes';
import { updateClass } from './modules/class';
import { updateStyle } from './modules/style';
import { updateEventListeners } from './modules/eventlisteners';
import { updateProps } from './modules/props';
export { VNode, VNodeData, vnode };
export { h } from './h';

type VNodeQueue = Array<VNode>;

const emptyNode = vnode('', {}, [], undefined, undefined);

function sameVnode(vnode1: VNode, vnode2: VNode): boolean {
  return vnode1.vkey === vnode2.vkey && vnode1.sel === vnode2.sel;
}

function isVnode(vnode: any): vnode is VNode {
  return vnode.sel !== undefined || vnode.elm !== undefined;
}

type KeyToIndexMap = {[key: string]: number};

function createKeyToOldIdx(children: Array<VNode>, beginIdx: number, endIdx: number): KeyToIndexMap {
  let i: number, map: KeyToIndexMap = {}, key: Key, ch;
  for (i = beginIdx; i <= endIdx; ++i) {
    ch = children[i];
    if (ch != null) {
      key = (<VNode>ch).vkey;
      if (key !== undefined) map[key] = i;
    }
  }
  return map;
}


export function Renderer(api: PlatformApi): RendererApi {

  function emptyNodeAt(elm: Element) {
    const id = elm.id ? '#' + elm.id : '';
    const c = elm.className ? '.' + elm.className.split(' ').join('.') : '';
    return vnode(api.$tagName(elm) + id + c, {}, [], undefined, elm);
  }

  function createRmCb(childElm: Node) {
    return function rmCb() {
      const parent = api.$parentNode(childElm);
      api.$removeChild(parent, childElm);
    };
  }

  function createElm(vnode: VNode, insertedVnodeQueue: VNodeQueue, parentElm: Node, hostContentNodes: HostContentNodes): Node {
    let i: any, data = vnode.vdata;
    if (data !== undefined) {
      if (isDef(i = data.hook) && isDef(i = i.init)) {
        i(vnode);
        data = vnode.vdata;
      }
    }
    let children = vnode.vchildren, sel = vnode.sel;
    if (sel === '!') {
      if (isUndef(vnode.vtext)) {
        vnode.vtext = '';
      }
      vnode.elm = api.$createComment(vnode.vtext as string);

    } else if (hostContentNodes && sel === 'slot') {
      // special case for manually relocating host content nodes
      // there their new home in either a named slot or the default slot
      let namedSlot = (data.attrs && data.attrs.name);
      let slotNodes: Node[];

      if (namedSlot) {
        slotNodes = hostContentNodes.$namedSlots && hostContentNodes.$namedSlots[namedSlot];

      } else {
        slotNodes = hostContentNodes.$defaultSlot;
      }

      if (slotNodes) {
        for (let nodeIndex = 0; nodeIndex < slotNodes.length; nodeIndex++) {
          // remove the host content node from it's original parent node
          (<ProxyElement>slotNodes[nodeIndex]).$tmpDisconnected = true;
          api.$removeChild(slotNodes[nodeIndex].parentNode, slotNodes[nodeIndex]);

          if (nodeIndex === slotNodes.length - 1) {
            // return the last node that gets appended
            // like any other Node that was created
            return slotNodes[nodeIndex];
          }

          // relocate the node to it's new home
          api.$appendChild(parentElm, slotNodes[nodeIndex]);
          delete (<ProxyElement>slotNodes[nodeIndex]).$tmpDisconnect;
        }
      }

    } else if (sel !== undefined) {
      // Parse selector
      const hashIdx = sel.indexOf('#');
      const dotIdx = sel.indexOf('.', hashIdx);
      const hash = hashIdx > 0 ? hashIdx : sel.length;
      const dot = dotIdx > 0 ? dotIdx : sel.length;
      const tag = hashIdx !== -1 || dotIdx !== -1 ? sel.slice(0, Math.min(hash, dot)) : sel;
      const elm = vnode.elm = isDef(data) && isDef(i = (data as VNodeData).ns) ? api.$createElementNS(i, tag)
                                                                               : api.$createElement(<any>tag);
      if (hash < dot) elm.setAttribute('id', sel.slice(hash + 1, dot));
      if (dotIdx > 0) elm.setAttribute('class', sel.slice(dot + 1).replace(/\./g, ' '));

      updateAttrs(emptyNode, vnode);
      updateClass(emptyNode, vnode);
      updateStyle(emptyNode, vnode);
      updateEventListeners(emptyNode, vnode);
      updateProps(emptyNode, vnode);

      data.ref && data.ref(elm);

      if (isArray(children)) {
        for (i = 0; i < children.length; ++i) {
          let ch = children[i];
          if (ch != null) {
            ch = <any>createElm(ch as VNode, insertedVnodeQueue, elm, hostContentNodes);
            if (ch) {
              api.$appendChild(elm, <any>ch);
            }
          }
        }
      } else if (isStringOrNumber(vnode.vtext)) {
        api.$appendChild(elm, api.$createTextNode(vnode.vtext));
      }
      i = (vnode.vdata as VNodeData).hook; // Reuse variable
      if (isDef(i)) {
        if (i.create) i.create(emptyNode, vnode);
        if (i.insert) insertedVnodeQueue.push(vnode);
      }
    } else {
      vnode.elm = api.$createTextNode(vnode.vtext as string);
    }
    return vnode.elm;
  }

  function addVnodes(parentElm: Node,
                     before: Node | null,
                     vnodes: Array<VNode>,
                     startIdx: number,
                     endIdx: number,
                     insertedVnodeQueue: VNodeQueue,
                     hostContentNodes: HostContentNodes) {
    for (; startIdx <= endIdx; ++startIdx) {
      const ch = vnodes[startIdx];
      if (ch != null) {
        api.$insertBefore(parentElm, createElm(ch, insertedVnodeQueue, parentElm, hostContentNodes), before);
      }
    }
  }

  function removeVnodes(parentElm: Node,
                        vnodes: Array<VNode>,
                        startIdx: number,
                        endIdx: number): void {
    for (; startIdx <= endIdx; ++startIdx) {
      let i: any, rm: () => void, ch = vnodes[startIdx];
      if (ch != null) {
        if (isDef(ch.sel)) {
          invokeDestroyHook(ch);
          rm = createRmCb(ch.elm as Node);
          if (isDef(i = ch.vdata) && isDef(i = i.hook) && isDef(i = i.remove)) {
            i(ch, rm);
          } else {
            rm();
          }
        } else { // Text node
          api.$removeChild(parentElm, ch.elm as Node);
        }
      }
    }
  }

  function updateChildren(parentElm: Node,
                          oldCh: Array<VNode>,
                          newCh: Array<VNode>,
                          insertedVnodeQueue: VNodeQueue,
                          hostContentNodes: HostContentNodes) {
    let oldStartIdx = 0, newStartIdx = 0;
    let oldEndIdx = oldCh.length - 1;
    let oldStartVnode = oldCh[0];
    let oldEndVnode = oldCh[oldEndIdx];
    let newEndIdx = newCh.length - 1;
    let newStartVnode = newCh[0];
    let newEndVnode = newCh[newEndIdx];
    let oldKeyToIdx: any;
    let idxInOld: number;
    let elmToMove: VNode;
    let before: any;

    while (oldStartIdx <= oldEndIdx && newStartIdx <= newEndIdx) {
      if (oldStartVnode == null) {
        oldStartVnode = oldCh[++oldStartIdx]; // Vnode might have been moved left
      } else if (oldEndVnode == null) {
        oldEndVnode = oldCh[--oldEndIdx];
      } else if (newStartVnode == null) {
        newStartVnode = newCh[++newStartIdx];
      } else if (newEndVnode == null) {
        newEndVnode = newCh[--newEndIdx];
      } else if (sameVnode(oldStartVnode, newStartVnode)) {
        patchVnode(oldStartVnode, newStartVnode, insertedVnodeQueue, hostContentNodes);
        oldStartVnode = oldCh[++oldStartIdx];
        newStartVnode = newCh[++newStartIdx];
      } else if (sameVnode(oldEndVnode, newEndVnode)) {
        patchVnode(oldEndVnode, newEndVnode, insertedVnodeQueue, hostContentNodes);
        oldEndVnode = oldCh[--oldEndIdx];
        newEndVnode = newCh[--newEndIdx];
      } else if (sameVnode(oldStartVnode, newEndVnode)) { // Vnode moved right
        patchVnode(oldStartVnode, newEndVnode, insertedVnodeQueue, hostContentNodes);
        api.$insertBefore(parentElm, oldStartVnode.elm as Node, api.$nextSibling(oldEndVnode.elm as Node));
        oldStartVnode = oldCh[++oldStartIdx];
        newEndVnode = newCh[--newEndIdx];
      } else if (sameVnode(oldEndVnode, newStartVnode)) { // Vnode moved left
        patchVnode(oldEndVnode, newStartVnode, insertedVnodeQueue, hostContentNodes);
        api.$insertBefore(parentElm, oldEndVnode.elm as Node, oldStartVnode.elm as Node);
        oldEndVnode = oldCh[--oldEndIdx];
        newStartVnode = newCh[++newStartIdx];
      } else {
        if (oldKeyToIdx === undefined) {
          oldKeyToIdx = createKeyToOldIdx(oldCh, oldStartIdx, oldEndIdx);
        }
        idxInOld = oldKeyToIdx[newStartVnode.vkey as string];
        if (isUndef(idxInOld)) { // New element
          api.$insertBefore(parentElm, createElm(newStartVnode, insertedVnodeQueue, parentElm, hostContentNodes), oldStartVnode.elm as Node);
          newStartVnode = newCh[++newStartIdx];
        } else {
          elmToMove = oldCh[idxInOld];
          if (elmToMove.sel !== newStartVnode.sel) {
            api.$insertBefore(parentElm, createElm(newStartVnode, insertedVnodeQueue, parentElm, hostContentNodes), oldStartVnode.elm as Node);
          } else {
            patchVnode(elmToMove, newStartVnode, insertedVnodeQueue, hostContentNodes);
            oldCh[idxInOld] = undefined as any;
            api.$insertBefore(parentElm, (elmToMove.elm as Node), oldStartVnode.elm as Node);
          }
          newStartVnode = newCh[++newStartIdx];
        }
      }
    }
    if (oldStartIdx > oldEndIdx) {
      before = newCh[newEndIdx + 1] == null ? null : newCh[newEndIdx + 1].elm;
      addVnodes(parentElm, before, newCh, newStartIdx, newEndIdx, insertedVnodeQueue, hostContentNodes);
    } else if (newStartIdx > newEndIdx) {
      removeVnodes(parentElm, oldCh, oldStartIdx, oldEndIdx);
    }
  }

  function patchVnode(oldVnode: VNode, vnode: VNode, insertedVnodeQueue: VNodeQueue, hostContentNodes: HostContentNodes) {
    if (!hostContentNodes && oldVnode.sel === 'slot') {
      return;
    }

    let i: any, hook: any;
    if (isDef(i = vnode.vdata) && isDef(hook = i.hook) && isDef(i = hook.prepatch)) {
      i(oldVnode, vnode);
    }

    const elm = vnode.elm = (oldVnode.elm as Node);
    let oldCh = oldVnode.vchildren;
    let ch = vnode.vchildren;
    if (oldVnode === vnode) return;
    if (vnode.vdata !== undefined) {
      updateAttrs(oldVnode, vnode);
      updateClass(oldVnode, vnode);
      updateStyle(emptyNode, vnode);
      updateEventListeners(oldVnode, vnode);
      updateProps(oldVnode, vnode);

      i = vnode.vdata.hook;
      if (isDef(i) && isDef(i = i.update)) i(oldVnode, vnode);
    }
    if (isUndef(vnode.vtext)) {
      if (isDef(oldCh) && isDef(ch)) {
        if (oldCh !== ch) {
          updateChildren((<HTMLElement>elm).shadowRoot || elm, oldCh as Array<VNode>, ch as Array<VNode>, insertedVnodeQueue, hostContentNodes);
        }

      } else if (isDef(ch)) {
        if (isDef(oldVnode.vtext)) api.$setTextContent(elm, '');
        addVnodes(elm, null, ch as Array<VNode>, 0, (ch as Array<VNode>).length - 1, insertedVnodeQueue, null);
      } else if (isDef(oldCh)) {
        removeVnodes(elm, oldCh as Array<VNode>, 0, (oldCh as Array<VNode>).length - 1);
      } else if (isDef(oldVnode.vtext)) {
        api.$setTextContent(elm, '');
      }
    } else if (oldVnode.vtext !== vnode.vtext) {
      api.$setTextContent(elm, vnode.vtext as string);
    }
    if (isDef(hook) && isDef(i = hook.postpatch)) {
      i(oldVnode, vnode);
    }
  }

  return function patch(oldVnode: VNode | Element, vnode: VNode, hostContentNodes?: HostContentNodes): VNode {
    let elm: Node, parent: Node;
    const insertedVnodeQueue: VNodeQueue = [];

    if (!isVnode(oldVnode)) {
      oldVnode = emptyNodeAt(oldVnode);
    }

    if (vnode.elm || sameVnode(oldVnode, vnode)) {
      patchVnode(oldVnode, vnode, insertedVnodeQueue, hostContentNodes);

    } else {
      elm = oldVnode.elm as Node;
      parent = api.$parentNode(elm);

      createElm(vnode, insertedVnodeQueue, parent, hostContentNodes);

      if (parent !== null) {
        api.$insertBefore(parent, vnode.elm as Node, api.$nextSibling(elm));
        removeVnodes(parent, [oldVnode], 0, 0);
      }
    }
    return vnode;
  };
}


export function invokeDestroyHook(vnode: VNode) {
  let i: any, j: number, data = vnode.vdata;
  if (data !== undefined) {
    updateEventListeners(vnode);
    if (vnode.vchildren !== undefined) {
      for (j = 0; j < vnode.vchildren.length; ++j) {
        i = vnode.vchildren[j];
        if (i != null && !isString(i)) {
          invokeDestroyHook(i);
        }
      }
    }
  }
}
