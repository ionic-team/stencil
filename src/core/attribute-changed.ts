import * as d from '../declarations';
import { toLowerCase } from '../util/helpers';


export function attributeChangedCallback(attrPropsMap: {[attr: string]: string}, elm: d.HostElement, attribName: string, newVal: string) {
  // look up to see if we have a property wired up to this attribute name
  const propName = attrPropsMap[toLowerCase(attribName)];
  if (propName) {
    // there is not need to cast the value since, it's already casted when
    // the prop is setted
    (elm as any)[propName] = newVal;
  }
}
