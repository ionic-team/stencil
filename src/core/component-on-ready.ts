import * as d from '../declarations';


export const initCoreComponentOnReady = (plt: d.PlatformApi, App: d.AppGlobal, win?: any, apps?: string[], queuedComponentOnReadys?: d.QueuedComponentOnReady[], i?: number) => {

  // add componentOnReady() to the App object
  // this also is used to know that the App's core is ready
  App.componentOnReady = (elm, resolve) => {

    if (!elm.nodeName.includes('-')) {
      resolve(null);
      return false;
    }

    const cmpMeta = plt.getComponentMeta(elm);
    if (cmpMeta) {
      if (plt.isCmpReady.has(elm)) {
        // element has already loaded, pass the resolve the element component
        // so we know that the resolve knows it this element is an app component
        resolve(elm);

      } else {
        // element hasn't loaded yet or it has an update in progress
        // add this resolve specifically to this elements on ready queue
        const onReadyCallbacks = plt.onReadyCallbacksMap.get(elm) || [];
        onReadyCallbacks.push(resolve);
        plt.onReadyCallbacksMap.set(elm, onReadyCallbacks);
      }
    }

    // return a boolean if this app recognized this element or not
    return !!cmpMeta;
  };

  if (queuedComponentOnReadys) {
    // we've got some componentOnReadys in the queue before the app was ready
    for (i = queuedComponentOnReadys.length - 1; i >= 0; i--) {
      // go through each element and see if this app recongizes it
      if (App.componentOnReady(queuedComponentOnReadys[i][0], queuedComponentOnReadys[i][1])) {
        // turns out this element belongs to this app
        // remove the resolve from the queue so in the end
        // all that's left in the queue are elements not apart of any apps
        queuedComponentOnReadys.splice(i, 1);
      }
    }

    for (i = 0; i < apps.length; i++) {
      if (!win[apps[i]].componentOnReady) {
        // there is at least 1 apps that isn't ready yet
        // so let's stop here cuz there's still app cores loading
        return;
      }
    }

    // if we got to this point then that means all of the apps are ready
    // and they would have removed any of their elements from queuedComponentOnReadys
    // so let's do the cleanup of the  remaining queuedComponentOnReadys
    for (i = 0; i < queuedComponentOnReadys.length; i++) {
      // resolve any queued componentsOnReadys that are left over
      // since these elements were not apart of any apps
      // call the resolve fn, but pass null so it's know this wasn't a known app component
      queuedComponentOnReadys[i][1](null);
    }
    queuedComponentOnReadys.length = 0;
  }
};
