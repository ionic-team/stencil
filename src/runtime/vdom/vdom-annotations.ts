import * as d from '../../declarations';
import { CONTENT_REF_ID, HYDRATE_CHILD_ID, HYDRATE_ID, NODE_TYPE, ORG_LOCATION_ID, SLOT_NODE_ID, TEXT_NODE_ID } from '../runtime-constants';
import { getHostRef } from '@platform';


export const insertVdomAnnotations = (doc: Document) => {
  if (doc != null) {
    const docData: DocData = {
      hostIds: 0,
      rootLevelIds: 0
    };
    const orgLocationNodes: d.RenderNode[] = [];

    parseVNodeAnnotations(doc, doc.body, docData, orgLocationNodes);

    orgLocationNodes.forEach(orgLocationNode => {
      if (orgLocationNode != null) {
        const nodeRef = orgLocationNode['s-nr'];

        let hostId = nodeRef['s-host-id'];
        let nodeId = nodeRef['s-node-id'];
        let childId = `${hostId}.${nodeId}`;

        if (hostId == null) {
          hostId = 0;
          docData.rootLevelIds++;
          nodeId = docData.rootLevelIds;
          childId = `${hostId}.${nodeId}`;

          if (nodeRef.nodeType === NODE_TYPE.ElementNode) {
            nodeRef.setAttribute(HYDRATE_CHILD_ID, childId);

          } else if (nodeRef.nodeType === NODE_TYPE.TextNode) {
            if (hostId === 0) {
              const textContent = nodeRef.nodeValue.trim();
              if (textContent === '') {
                // useless whitespace node at the document root
                orgLocationNode.remove();
                return;
              }
            }
            const commentBeforeTextNode = doc.createComment(childId);
            commentBeforeTextNode.nodeValue = `${TEXT_NODE_ID}.${childId}`;
            nodeRef.parentNode.insertBefore(commentBeforeTextNode, nodeRef);
          }
        }

        let orgLocationNodeId = `${ORG_LOCATION_ID}.${childId}`;

        const orgLocationParentNode = orgLocationNode.parentElement as d.RenderNode;
        if (orgLocationParentNode && orgLocationParentNode['s-sd']) {
          // ending with a . means that the parent element
          // of this node's original location is a shadow dom element
          // and this node is apart of the root level light dom
          orgLocationNodeId += `.`;
        }

        orgLocationNode.nodeValue = orgLocationNodeId;
      }
    });
  }
};


const parseVNodeAnnotations = (doc: Document, node: d.RenderNode, docData: DocData, orgLocationNodes: d.RenderNode[]) => {
  if (node == null) {
    return;
  }

  if (node['s-nr'] != null) {
    orgLocationNodes.push(node);
  }

  if (node.nodeType === NODE_TYPE.ElementNode) {
    node.childNodes.forEach(childNode => {
      const hostRef = getHostRef(childNode);
      if (hostRef != null) {
        const cmpData: CmpData = {
          nodeIds: 0
        };
        insertVNodeAnnotations(doc, childNode as any, hostRef.$vnode$, docData, cmpData);
      }

      parseVNodeAnnotations(doc, childNode as any, docData, orgLocationNodes);
    });
  }
};


const insertVNodeAnnotations = (doc: Document, hostElm: d.HostElement, vnode: d.VNode, docData: DocData, cmpData: CmpData) => {
  if (vnode != null) {
    const hostId = ++docData.hostIds;

    hostElm.setAttribute(HYDRATE_ID, hostId as any);

    if (hostElm['s-cr'] != null) {
      hostElm['s-cr'].nodeValue = `${CONTENT_REF_ID}.${hostId}`;
    }

    if (vnode.$children$ != null) {
      const depth = 0;
      vnode.$children$.forEach((vnodeChild, index) => {
        insertChildVNodeAnnotations(doc, vnodeChild, cmpData, hostId, depth, index);
      });
    }
  }
};


const insertChildVNodeAnnotations = (doc: Document, vnodeChild: d.VNode, cmpData: CmpData, hostId: number, depth: number, index: number) => {
  const childElm = vnodeChild.$elm$ as d.RenderNode;
  if (childElm == null) {
    return;
  }

  const nodeId = cmpData.nodeIds++;
  const childId = `${hostId}.${nodeId}.${depth}.${index}`;

  childElm['s-host-id'] = hostId;
  childElm['s-node-id'] = nodeId;

  if (childElm.nodeType === NODE_TYPE.ElementNode) {
    childElm.setAttribute(HYDRATE_CHILD_ID, childId);

  } else if (childElm.nodeType === NODE_TYPE.TextNode) {
    const parentNode = childElm.parentNode;
    if (parentNode.nodeName !== 'STYLE') {
      const textNodeId = `${TEXT_NODE_ID}.${childId}`;

      const commentBeforeTextNode = doc.createComment(textNodeId);
      parentNode.insertBefore(commentBeforeTextNode, childElm);
    }

  } else if (childElm.nodeType === NODE_TYPE.CommentNode) {
    if (childElm['s-sr']) {
      const slotName = (childElm['s-sn'] || '');
      const slotNodeId = `${SLOT_NODE_ID}.${childId}.${slotName}`;
      childElm.nodeValue = slotNodeId;
    }
  }

  if (vnodeChild.$children$ != null) {
    const childDepth = depth + 1;
    vnodeChild.$children$.forEach((vnode, index) => {
      insertChildVNodeAnnotations(doc, vnode, cmpData, hostId, childDepth, index);
    });
  }
};


interface DocData {
  hostIds: number;
  rootLevelIds: number;
}


interface CmpData {
  nodeIds: number;
}
