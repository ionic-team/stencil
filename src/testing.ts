export { register, render } from './testing/index';
