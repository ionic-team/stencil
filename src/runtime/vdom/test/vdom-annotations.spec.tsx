import { Component, h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';

import { insertVdomAnnotations } from '../vdom-annotations';

describe('vdom-annotations', () => {
  let root: HTMLElement;

  beforeEach(async () => {
    @Component({ tag: 'component-a-test', scoped: true })
    class ComponentA {
      render() {
        return (
          <div>
            <slot></slot>
          </div>
        );
      }
    }

    @Component({ tag: 'component-b-test', scoped: true })
    class ComponentB {
      render() {
        return (
          <div>
            <slot></slot>
          </div>
        );
      }
    }

    const { root: rootElm } = await newSpecPage({
      components: [ComponentA, ComponentB],
      html: `<section>
        <component-a-test>slotContent</component-a-test>
        <component-b-test>slotContent</component-b-test>
      </section>`,
    });
    root = rootElm;
  });

  it('should add annotations when no static component is given', () => {
    insertVdomAnnotations(root.ownerDocument, []);
    /**
      <section>
        <component-a-test s-id="1">
            <!--r.1-->o.0.1
            <div c-id="1.0.0.0">
              <!--t.1.1.1.0--> <!--t.0.1-->slotContent
            </div>
        </component-a-test>
        <component-b-test s-id="2">
            <!--r.2-->o.0.2
            <div c-id="2.0.0.0">
              <!--t.2.1.1.0--> <!--t.0.2-->slotContent
            </div>
        </component-b-test>
      </section>
     */
    expect(root.ownerDocument.body.innerHTML).toMatchSnapshot();
  });

  it('should add annotations when component-a-test is given as static component', () => {
    insertVdomAnnotations(root.ownerDocument, ['component-a-test']);
    /**
      <section>
        <component-a-test>
            <!---->o.0.1
            <div>
              <!--t.0.1-->slotContent
            </div>
        </component-a-test>
        <component-b-test s-id="1">
            <!--r.1-->o.0.2
            <div c-id="1.0.0.0">
              <!--t.1.1.1.0--> <!--t.0.2-->slotContent
            </div>
        </component-b-test>
      </section>
     */
    expect(root.ownerDocument.body.innerHTML).toMatchSnapshot();
  });

  it('should add annotations when component-a-test and component-b-test is given as static component', () => {
    insertVdomAnnotations(root.ownerDocument, ['component-a-test', 'component-b-test']);
    /**
      <section>
        <component-a-test>
            <!---->o.0.1
            <div>
              <!--t.0.1-->slotContent
            </div>
        </component-a-test>
        <component-b-test>
            <!---->o.0.2
            <div>
              <!--t.0.2-->slotContent
            </div>
        </component-b-test>
      </section>
     */
    expect(root.ownerDocument.body.innerHTML).toMatchSnapshot();
  });
});
