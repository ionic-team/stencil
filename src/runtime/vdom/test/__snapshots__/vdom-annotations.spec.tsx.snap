// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vdom-annotations should add annotations when component-a-test and component-b-test is given as static component 1`] = `"<section> <component-a-test><!---->o.0.1<div> <!--t.0.1-->slotContent</div></component-a-test> <component-b-test><!---->o.0.2<div> <!--t.0.2-->slotContent</div></component-b-test> </section>"`;

exports[`vdom-annotations should add annotations when component-a-test is given as static component 1`] = `"<section> <component-a-test><!---->o.0.1<div> <!--t.0.1-->slotContent</div></component-a-test> <component-b-test s-id=\\"1\\"><!--r.1-->o.0.2<div c-id=\\"1.0.0.0\\"><!--t.1.1.1.0--> <!--t.0.2-->slotContent</div></component-b-test> </section>"`;

exports[`vdom-annotations should add annotations when no static component is given 1`] = `"<section> <component-a-test s-id=\\"1\\"><!--r.1-->o.0.1<div c-id=\\"1.0.0.0\\"><!--t.1.1.1.0--> <!--t.0.1-->slotContent</div></component-a-test> <component-b-test s-id=\\"2\\"><!--r.2-->o.0.2<div c-id=\\"2.0.0.0\\"><!--t.2.1.1.0--> <!--t.0.2-->slotContent</div></component-b-test> </section>"`;
